### DISCUSSION TITLE
Password recovery from firmware build

### DISCUSSION THREAD

**Discussion by shaman79** [arendst/Tasmota#21797] [Support request]
Hi there, so it happened that I flashed my device with firmware build that had web password set using #WEB_PASSWORD variable during build.
Unfortunately I forgot the password and I do not have the used user_override.h file, but I have the firmware build.

Is there a way to extract the web admin password from the firmware file?

**Comment by HofeBY** [comment]
Try from browser with active Device

http://DeviceIPv4/ufs/.settings

if downloded file arrived, look with text editor.

**Reply by shaman79** [reply]
I got:

File not found

URI: /ufs/.settings
Method: GET
Arguments: 0

**Reply by HofeBY** [reply]
you substitute "DeviceIPv4" with the IP Adress of the Device ?

Maybe File download not included in your Binary or Webpassword is blocking download, as I expect.

**Reply by shaman79** [reply]
Sure, the IP is 192.168.4.1 as the device is in defaults. 
I was looking into the file and for example MQTT password is there in plaintext. I was just hoping that web password is there in plintext too...seems not to be the case.

**Reply by sfromis** [reply]
On devices with a file system (not the case for nearly all ESP8266 devices), the plausible URL would be `http://tasmotadevice/ufsd?download=/.settings` which would not give you a readable file, but the tool decode-config would be the thing to try. **However**, this does not work without you knowing the `WebPassword`, as that would be just too bad security.

The simple way would be to set a new `WebPassword` via MQTT (if you have that set up) or serial.

**Reply by sfromis** [reply]
> I was looking into the file

What exact file do you mean?

If you mean the binary you flashed, the password is likely there, but not easy to spot if you do not know it already.... And if you have flashed via serial, just setup a serial console to change the password.

Anyway, the straightforward strategy would be to build again with a known web password.

**Reply by HofeBY** [reply]
Just locking one of my Devices (SML reader) by activating Webpassword.
Then open http://DeviceIPv4/ufs/.settings in Browser
get asked Login: admin password:Idefind
then download started and I take a look into File & found this webpassword as plain in text.
Between "Device Name" and "IP"

As I expect, download only with password possible.

**Reply by sfromis** [reply]
Most devices does not have that web server included in the build.

**Reply by HofeBY** [reply]
@sfromis
ESP32 with filesystem (320kByte) on Chip
file .settings downloaded via http://DeviceIPv4/ufs/.settings
If you inspect by filesystem tools of tasmota, you get shown nonens !
Download and inspect on Windows or Linux.
![ESP32Files](https://github.com/user-attachments/assets/764c0147-389e-45e3-bb90-094ae994c3a0)

**Reply by sfromis** [reply]
Sure, that worked for you, but it is **very** common that people are using ESP8266 devices without such details available. That's why I posted a UFSD link as an option with higher chance of success, if there even is a file system (which is not on most ESP8266 devices).

Hence my "most devices", where your suggestions just cannot work.

Even with a standard precompiled ESP32 build, your suggestion did not work in a quick test....

**Reply by HofeBY** [reply]
@sfromis
It was a try to get the password on an easy way.
shaman79 has it´s used firmware.bin and if not it is possible to download it from the physical device.
than it is possible to inspect it by a Hexeditor.

**Reply by shaman79** [reply]
Yep, the only thing I have is the .bin file. Web password is probably hashed or somehow encoded during build. It is not visible anywhere in the bin file.

**Reply by sfromis** [reply]
No, it is in plain text in the binary, and very visible, as long as you are not looking at a compressed `.gz` file.

**Reply by Jason2866** [reply]
exactly, Tasmota firmware and all settings are not encrypted. The devices needs to be secured against unwanted people access of the hardware.

**Comment by HofeBY** [comment]
I don't want to emphasize it too much.
But even if you protect web access with a password, the authentication is done via http and can therefore be read by the man in the middle in promiscuous mode.

**Reply by sfromis** [reply]
Right, it is not anywhere "real security". Of course, if you already have intruders on your network, you may have a lot worse problems to deal with than a Tasmota device.
